### PR TITLE
Translate to Simplified Chinese: How does React tell a class from a function

### PR DESCRIPTION
**Pull request by tonghuashuo**
This pull request contains a translation of the article "How does React tell a class from a function" to Simplified Chinese.

### CODE REVIEW

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+// React 内部
+const instance = new Greeting(props); // Greeting {}
+const result = instance.render(); // <p>Hello</p>
+```
+
+无论哪种情况 React 的目标都是去获取渲染后的节点（在这个案例中，`<p>Hello</p>`）。但具体的步骤取决于 `Greeting` 是如何定义的。
+
+**所以 React 是怎么知道某样东西是 class 还是 function 的呢？**
+
+就像我 [上一篇博客](https://overreacted.io/why-do-we-write-super-props/) 中提到的，**你并不 *需要* 知道这个才能高效使用 React。** 我多年来都不知道这个。请不要把这变成一道面试题。事实上，这篇博客更多的是关于 JavaScript 而不是 React。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
就像我 [上一篇博客](https://overreacted.io/why-do-we-write-super-props/) 中提到的，**你并不*需要*知道这个才能高效使用 React。** 我多年来都不知道这个。请不要把这变成一道面试题。事实上，这篇博客更多的是关于 JavaScript 而不是 React。
```
I removed unnecessary whitespaces.

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+// 如果 Greeting 是一个类
+const instance = new Greeting(props); // Greeting {}
+const result = instance.render(); // <p>Hello</p>
+```
+
+我们来简单看一下 `new` 在 JavaScript 是干什么的。
+
+* * *
+
+在过去，JavaScript 还没有类。但是，你可以使用普通函数来模拟。**具体来讲，只要在函数调用前加上 `new` 操作符，你就可以把 *任何* 函数当做一个类的构造函数来用：**
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
在过去，JavaScript 还没有类。但是，你可以使用普通函数来模拟。**具体来讲，只要在函数调用前加上 `new` 操作符，你就可以把*任何*函数当做一个类的构造函数来用：**
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+  this.name = name;
+}
+
+var fred = new Person('Fred'); // ✅ Person {name: 'Fred'}
+var george = Person('George'); // 🔴 没用的
+```
+
+现在你依然可以这样写！在 DevTools 里试试吧。
+
+如果你调用 `Person('Fred')` 时 **没有** 加 `new`，其中的 `this` 会指向某个全局且无用的东西（比如，`window` 或者 `undefined`），因此我们的代码会崩溃，或者做一些像设置 `window.name` 之类的傻事。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
如果你调用 `Person('Fred')` 时**没有**加 `new`，其中的 `this` 会指向某个全局且无用的东西（比如，`window` 或者 `undefined`），因此我们的代码会崩溃，或者做一些像设置 `window.name` 之类的傻事。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+
+通过在调用前增加 `new`，我们说：“嘿 JavaScript，我知道 `Person` 只是个函数，但让我们假装它是个构造函数吧。**创建一个 `{}` 对象并把 `Person` 中的 `this` 指向那个对象，以便我可以通过类似 `this.name` 的形式去设置一些东西，然后把这个对象返回给我。**”
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
通过在调用前增加 `new`，我们说：「嘿 JavaScript，我知道 `Person` 只是个函数，但让我们假装它是个构造函数吧。**创建一个 `{}` 对象并把 `Person` 中的 `this` 指向那个对象，以便我可以通过类似 `this.name` 的形式去设置一些东西，然后把这个对象返回给我。**」
```
I suggest using full-width corner brackets `「` and `」`. See below for reference. 
https://github.com/mzlogin/chinese-copywriting-guidelines#%E5%85%A8%E8%A7%92%E5%92%8C%E5%8D%8A%E8%A7%92

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+  sayHi() {
+    alert('Hi, I am ' + this.name);
+  }
+}
+
+let fred = new Person('Fred');
+fred.sayHi();
+```
+
+***捕捉开发者的本意*** 是语言和 API 设计中非常重要的一点。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
***捕捉开发者的本意***是语言和 API 设计中非常重要的一点。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+
+如果你写了一个函数，JavaScript 没办法判断它应该像 `alert()` 一样被调用，还是应该被视作像 `new Person()` 一样的构造函数。忘记给像 `Person` 这样的函数指定 `new` 会导致令人费解的行为。
+
+**类语法允许我们说：“这不仅仅是个函数 —— 这是个类并且它有构造函数”。** 如果你在调用它时忘了加 `new`，JavaScript 会报错：
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
**类语法允许我们说：「这不仅仅是个函数 —— 这是个类并且它有构造函数。」**如果你在调用它时忘了加 `new`，JavaScript 会报错：
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+  }
+  // Our code:
+  this.name = name;
+}
+
+new Person('Fred'); // ✅ OK
+Person('George');   // 🔴 无法把类当做函数来调用
+```
+
+你或许已经在你构建出来的包中见过类似的代码，这就是那些 `_classCallCheck` 函数做的事。（你可以通过启用“loose mode”来关闭检查以减小构建包的尺寸，但这或许会使你最终转向真正的原生类时变得复杂）
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
你或许已经在你构建出来的包中见过类似的代码，这就是那些 `_classCallCheck` 函数做的事。（你可以通过启用「loose mode」来关闭检查以减小构建包的尺寸，但这或许会使你最终转向真正的原生类时变得复杂。）
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+* * *
+
+至此，你应该已经大致理解了调用时加不加 `new` 的差别：
+
+|            | `new Person()`               | `Person()`                          |
+| ---------- | ---------------------------- | ----------------------------------- |
+| `class`    | ✅ `this` 是一个 `Person` 实例 | 🔴 `TypeError`                      |
+| `function` | ✅ `this` 是一个 `Person` 实例 | 😳 `this` 是 `window` 或 `undefined` |
+
+这就是 React 正确调用你的组件很重要的原因。 **如果你的组件被定义为一个类，React 需要使用 `new` 来调用它**
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
这就是 React 正确调用你的组件很重要的原因。**如果你的组件被定义为一个类，React 需要使用 `new` 来调用它。**
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+
+所以 React 能检查出某样东西是否是类吗？
+
+没那么容易！即便我们能够 [在 JavaScript 中区分类和函数](https://stackoverflow.com/questions/29093396/how-do-you-check-the-difference-between-an-ecmascript-6-class-and-function)，面对被 Babel 等工具处理过的类这还是没用。对浏览器而言，它们只是不同的函数。这是 React 的不幸。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
没那么容易！即便我们能够[在 JavaScript 中区分类和函数](https://stackoverflow.com/questions/29093396/how-do-you-check-the-difference-between-an-ecmascript-6-class-and-function)，面对被 Babel 等工具处理过的类这还是没用。对浏览器而言，它们只是不同的函数。这是 React 的不幸。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+对于常规函数，用 `new` 调用会给它们一个 `this` 作为对象实例。对于用作构造函数的函数（比如我们前面提到的 `Person`）是可取的，但对函数组件这或许就比较令人困惑了：
+
+```jsx
+function Greeting() {
+  // 我们并不期望 `this` 在这里表示任何类型的实例
+  return <p>Hello</p>;
+}
+```
+
+这暂且还能忍，还有两个 *其他* 理由会扼杀这个想法。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
这暂且还能忍，还有两个*其他*理由会扼杀这个想法。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 11]
```diff
@@ -0,0 +1,549 @@
...
+* * *
+
+关于为什么总是使用 `new` 是没用的的第一个理由是，对于原生的箭头函数（不是那些被 Babel 编译过的），用 `new` 调用会抛出一个错误：
+
+```jsx
+const Greeting = () => <p>Hello</p>;
+new Greeting(); // 🔴 Greeting 不是一个构造函数
+```
+
+这个行为是遵循箭头函数的设计而刻意为之的。箭头函数的一个附带作用是它 *没有* 自己的 `this` 值 —— `this` 解析自离得最近的常规函数：
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
这个行为是遵循箭头函数的设计而刻意为之的。箭头函数的一个附带作用是它*没有*自己的 `this` 值 —— `this` 解析自离得最近的常规函数：
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 12]
```diff
@@ -0,0 +1,549 @@
...
+        size={this.props.size}
+        name={friend.name}
+        key={friend.id}
+      />
+    );
+  }
+}
+```
+
+OK，所以 **箭头函数没有自己的 `this`。**但这意味着它作为构造函数是完全无用的！
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
OK，所以**箭头函数没有自己的 `this`。**但这意味着它作为构造函数是完全无用的！
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 13]
```diff
@@ -0,0 +1,549 @@
...
+
+```jsx
+const Person = (name) => {
+  // 🔴 这么写是没有意义的！
+  this.name = name;
+}
+```
+
+因此，**JavaScript 不允许用 `new` 调用箭头函数。** 如果你这么做，你或许已经犯了错，最好早点告诉你。这和 JavaScript 不让你 *不加* `new` 去调用一个类是类似的。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
因此，**JavaScript 不允许用 `new` 调用箭头函数。** 如果你这么做，你或许已经犯了错，最好早点告诉你。这和 JavaScript 不让你*不加* `new` 去调用一个类是类似的。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 14]
```diff
@@ -0,0 +1,549 @@
...
+
+这样很不错，但这也让我们的计划受阻。React 不能简单对所有东西都使用 `new`，因为会破坏箭头函数！我们可以利用箭头函数没有 `prototype` 的特点来检测箭头函数，不对它们使用 `new`：
+
+```jsx
+(() => {}).prototype // undefined
+(function() {}).prototype // {constructor: f}
+```
+
+但这对于被 Babel 编译过的函数是 [没用](https://github.com/facebook/react/issues/4599#issuecomment-136562930) 的。这或许没什么大不了，但还有另一个原因使得这条路不会有结果。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
但这对于被 Babel 编译过的函数是[没用](https://github.com/facebook/react/issues/4599#issuecomment-136562930) 的。这或许没什么大不了，但还有另一个原因使得这条路不会有结果。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 15]
```diff
@@ -0,0 +1,549 @@
...
+```jsx
+function Greeting() {
+  return 'Hello';
+}
+
+Greeting(); // ✅ 'Hello'
+new Greeting(); // 😳 Greeting {}
+```
+
+这，再一次，和 [`new` 操作符](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new) 的怪异设计有关。如我们之前所看到的，`new` 告诉 JavaScript 引擎去创建一个对象，让这个对象成为函数内部的 `this`，然后把这个对象作为 `new` 的结果给我们。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
这，再一次，和[`new` 操作符](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new)的怪异设计有关。如我们之前所看到的，`new` 告诉 JavaScript 引擎去创建一个对象，让这个对象成为函数内部的 `this`，然后把这个对象作为 `new` 的结果给我们。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 16]
```diff
@@ -0,0 +1,549 @@
...
+
+然而，JavaScript 也允许一个使用 `new` 调用的函数返回另一个对象以 **覆盖** `new` 的返回值。或许，这在我们利用诸如“对象池模式”来对组件进行复用时是被认为有用的：
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
然而，JavaScript 也允许一个使用 `new` 调用的函数返回另一个对象以**覆盖** `new` 的返回值。或许，这在我们利用诸如“对象池模式”来对组件进行复用时是被认为有用的：
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 17]
```diff
@@ -0,0 +1,549 @@
...
+  this.x = x;
+  this.y = y;
+}
+
+var a = new Vector(1, 1);
+var b = new Vector(0, 0);
+var c = new Vector(0, 0); // 😲 b === c
+```
+
+然而，如果一个函数的返回值 *不是* 一个对象，它会被 `new` *完全忽略*。如果你返回了一个字符串或数字，就好像完全没有 `return` 一样。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
然而，如果一个函数的返回值*不是*一个对象，它会被 `new` *完全忽略*。如果你返回了一个字符串或数字，就好像完全没有 `return` 一样。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 18]
```diff
@@ -0,0 +1,549 @@
...
+new Answer(); // 😳 Answer {}
+```
+
+当使用 `new` 调用函数时，是没办法读取原始类型（例如一个数字或字符串）的返回值的。因此如果 React 总是使用 `new`，就没办法增加对返回字符串的组件的支持！
+
+这是不可接受的，因此我们必须妥协。
+
+* * *
+
+至此我们学到了什么？React 在调用类（包括 Babel 输出的）时 *需要用* `new`，但在调用常规函数或箭头函数时（包括 Babel 输出的）**不需要用** `new`，并且没有可靠的方法来区分这些情况。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
至此我们学到了什么？React 在调用类（包括 Babel 输出的）时*需要用* `new`，但在调用常规函数或箭头函数时（包括 Babel 输出的）**不需要用** `new`，并且没有可靠的方法来区分这些情况。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 19]
```diff
@@ -0,0 +1,549 @@
...
+```jsx
+class A {}
+class B extends A {}
+
+console.log(B.prototype instanceof A); // true
+```
+
+我知道你在想什么，刚才发生了什么？！为了回答这个问题，我们需要理解 JavaScript 原型。
+
+你或许对“原型链”很熟悉。JavaScript 中的每一个对象都有一个“原型”。当我们写 `fred.sayHi()` 但 `fred` 对象没有 `sayHi` 属性，我们尝试到 `fred` 的原型上去找 `sayHi` 属性。要是我们在这儿找不到，就去找原型链的下一个原型 —— `fred` 的原型的原型，以此类推。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
你或许对「原型链」很熟悉。JavaScript 中的每一个对象都有一个「原型」。当我们写 `fred.sayHi()` 但 `fred` 对象没有 `sayHi` 属性，我们尝试到 `fred` 的原型上去找 `sayHi` 属性。要是我们在这儿找不到，就去找原型链的下一个原型 —— `fred` 的原型的原型，以此类推。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 20]
```diff
@@ -0,0 +1,549 @@
...
+
+**费解的是，一个类或函数的 `prototype` 属性 _并不_ 指向那个值的原型。** 我没开玩笑。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
**费解的是，一个类或函数的 `prototype` 属性 _并不_ 指向那个值的原型。**我没开玩笑。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 21]
```diff
@@ -0,0 +1,549 @@
...
+
+```jsx
+function Person() {}
+
+console.log(Person.prototype); // 🤪 不是 Person 的原型
+console.log(Person.__proto__); // 😳 Person 的原型
+```
+
+因此“原型链”更像是 `__proto__.__proto__.__proto__` 而不是 `prototype.prototype.prototype`，我花了好几年才搞懂这一点。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
因此「原型链」更像是 `__proto__.__proto__.__proto__` 而不是 `prototype.prototype.prototype`，我花了好几年才搞懂这一点。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 22]
```diff
@@ -0,0 +1,549 @@
...
+
+那么函数和类的 `prototype` 属性又是什么？**是用 `new` 调用那个类或函数生成的所有对象的 `__proto__` ！**
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
那么函数和类的 `prototype` 属性又是什么？**是用 `new` 调用那个类或函数生成的所有对象的 `__proto__`！**
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 23]
```diff
@@ -0,0 +1,549 @@
...
+// 1. fred 有 sayHi 属性吗？不。
+// 2. fred.__proto__ 有 sayHi 属性吗？是的，调用它！
+
+fred.toString();
+// 1. fred 有 toString 属性吗？不。
+// 2. fred.__proto__ 有 toString 属性吗？不。
+// 3. fred.__proto__.__proto__ 有 toString 属性吗？是的，调用它！
+```
+
+在实战中，你应该几乎永远不需要直接在代码里动到 `__proto__` 除非你在调试和原型链相关的问题。如果你想让某样东西在 `fred.__proto__` 上可用，你应该把它放在 `Person.prototype`，至少它最初是这么设计的。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
在实战中，你应该几乎永远不需要直接在代码里动到 `__proto__` ，除非你在调试和原型链相关的问题。如果你想让某样东西在 `fred.__proto__` 上可用，你应该把它放在 `Person.prototype`，至少它最初是这么设计的。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 24]
```diff
@@ -0,0 +1,549 @@
...
+
+`__proto__` 属性甚至一开始就不应该被浏览器暴露出来，因为原型链应该被视为一个内部概念，然而某些浏览器增加了 `__proto__` 并最终勉强被标准化（但已被废弃并推荐使用 `Object.getPrototypeOf()`）。
+
+**然而一个名叫“原型”的属性却给不了我一个值的“原型”这一点还是很让我困惑**（例如，`fred.prototype` 是未定义的，因为 `fred` 不是一个函数）。个人观点，我觉得这是即便有经验的开发者也容易误解 JavaScript 原型链的最大原因。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
**然而一个名叫「原型」的属性却给不了我一个值的「原型」这一点还是很让我困惑**（例如，`fred.prototype` 是未定义的，因为 `fred` 不是一个函数）。个人观点，我觉得这是即便有经验的开发者也容易误解 JavaScript 原型链的最大原因。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 25]
```diff
@@ -0,0 +1,549 @@
...
+console.log(c.__proto__); // Greeting.prototype
+console.log(c.__proto__.__proto__); // React.Component.prototype
+console.log(c.__proto__.__proto__.__proto__); // Object.prototype
+
+c.render();      // 在 c.__proto__ (Greeting.prototype) 上找到
+c.setState();    // 在 c.__proto__.__proto__ (React.Component.prototype) 上找到
+c.toString();    // 在 c.__proto__.__proto__.__proto__ (Object.prototype) 上找到
+```
+
+换句话说，**当你在使用类的时候，实例的 `__proto__` 链“镜像”了类的层级结构：**
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
换句话说，**当你在使用类的时候，实例的 `__proto__` 链「镜像」了类的层级结构：**
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 26]
```diff
@@ -0,0 +1,549 @@
...
+//       .__proto__ → Object.prototype (✅ 找到了！)
+
+console.log(greeting instanceof Banana); // false
+// greeting (🕵️‍ 我们从这儿开始)
+//   .__proto__ → Greeting.prototype
+//     .__proto__ → React.Component.prototype
+//       .__proto__ → Object.prototype (🙅‍ 没找到！)
+```
+
+但这用来判断一个类是否扩展了另一个类还是有效的
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
但这用来判断一个类是否扩展了另一个类还是有效的：
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 27]
```diff
@@ -0,0 +1,549 @@
...
+//   .__proto__ → Greeting.prototype (🕵️‍ 我们从这儿开始)
+//     .__proto__ → React.Component.prototype (✅ 找到了！)
+//       .__proto__ → Object.prototype
+```
+
+这种检查方式就是我们判断某样东西是一个 React 组件类还是一个常规函数的方式。
+
+* * *
+
+然而 React 并不是这么做的 😳
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
然而 React 并不是这么做的。😳
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 28]
```diff
@@ -0,0 +1,549 @@
...
+
+关于 `instanceof` 解决方案有一点附加说明，当页面上有多个 React 副本，并且我们要检查的组件继承自 **另一个** React 副本的 `React.Component` 时，这种方法是无效的。在一个项目里混合多个 React 副本是不好的，原因有很多，但站在历史角度来看，我们试图尽可能避免问题。（有了 Hooks，我们 [或许得](https://github.com/facebook/react/issues/13991) 强制避免重复）
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
关于 `instanceof` 解决方案有一点附加说明，当页面上有多个 React 副本，并且我们要检查的组件继承自 **另一个** React 副本的 `React.Component` 时，这种方法是无效的。在一个项目里混合多个 React 副本是不好的，原因有很多，但站在历史角度来看，我们试图尽可能避免问题。（有了 Hooks，我们[或许得](https://github.com/facebook/react/issues/13991)强制避免重复）
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 29]
```diff
@@ -0,0 +1,549 @@
...
+
+另一点启发可以是去检查原型链上的 `render` 方法。然而，当时还 [不确定](https://github.com/facebook/react/issues/4599#issuecomment-129714112) 组件的 API 会如何演化。每一次检查都有成本，所以我们不想再多加了。如果 `render` 被定义为一个实例方法，例如使用类属性语法，这个方法也会失效。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
另一点启发可以是去检查原型链上的 `render` 方法。然而，当时还[不确定](https://github.com/facebook/react/issues/4599#issuecomment-129714112)组件的 API 会如何演化。每一次检查都有成本，所以我们不想再多加了。如果 `render` 被定义为一个实例方法，例如使用类属性语法，这个方法也会失效。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 30]
```diff
@@ -0,0 +1,549 @@
...
+
+因此, React 为基类 [增加了](https://github.com/facebook/react/pull/4663) 一个特别的标记。React 检查是否有这个标记，以此知道某样东西是否是一个 React 组件类。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
因此, React 为基类[增加了](https://github.com/facebook/react/pull/4663) 一个特别的标记。React 检查是否有这个标记，以此知道某样东西是否是一个 React 组件类。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 31]
```diff
@@ -0,0 +1,549 @@
...
+// React 内部
+class Component {}
+Component.isReactClass = {};
+
+// 我们可以像这样检查它
+class Greeting extends Component {}
+console.log(Greeting.isReactClass); // ✅ 是的
+```
+
+然而，有些我们希望作为目标的类实现 [并没有](https://github.com/scala-js/scala-js/issues/1900) 复制静态属性（或设置非标准的 `__proto__`），标记也因此丢失。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
然而，有些我们希望作为目标的类实现[并没有](https://github.com/scala-js/scala-js/issues/1900) 复制静态属性（或设置非标准的 `__proto__`），标记也因此丢失。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 32]
```diff
@@ -0,0 +1,549 @@
...
+
+这也是为什么 React 把这个标记 [移动到了](https://github.com/facebook/react/pull/5021) `React.Component.prototype`：
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
这也是为什么 React 把这个标记[移动到了](https://github.com/facebook/react/pull/5021) `React.Component.prototype`：
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 33]
```diff
@@ -0,0 +1,549 @@
...
+console.log(Greeting.prototype.isReactComponent); // ✅ 是的
+```
+
+**说真的这就是全部了。**
+
+你或许奇怪为什么是一个对象而不是一个布尔值。实战中这并不重要，但早期版本的 Jest（在 Jest 商品化之前）是默认开始自动模拟功能的，生成的模拟数据省略掉了原始类型属性，[破坏了检查](https://github.com/facebook/react/pull/4663#issuecomment-136533373)。谢了，Jest。
+
+一直到今天，[React 都在用](https://github.com/facebook/react/blob/769b1f270e1251d9dbdce0fcbd9e92e502d059b8/packages/react-reconciler/src/ReactFiber.js#L297-L300) `isReactComponent` 进行检查。
+
+如果你不扩展 `React.Component`，React 不会在原型上找到 `isReactComponent`，因此就不会把组件当做类处理。现在你知道为什么解决 `Cannot call a class as a function` 错误的 [得票数最高的答案](https://stackoverflow.com/a/42680526/458193) 是增加 `extends React.Component`。最后，我们还 [增加了一项警告](https://github.com/facebook/react/pull/11168)，当 `prototype.render` 存在但 `prototype.isReactComponent` 不存在时会发出警告。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
如果你不扩展 `React.Component`，React 不会在原型上找到 `isReactComponent`，因此就不会把组件当做类处理。现在你知道为什么解决 `Cannot call a class as a function` 错误的[得票数最高的答案](https://stackoverflow.com/a/42680526/458193)是增加 `extends React.Component`。最后，我们还[增加了一项警告](https://github.com/facebook/react/pull/11168)，当 `prototype.render` 存在但 `prototype.isReactComponent` 不存在时会发出警告。
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 34]
```diff
@@ -0,0 +1,549 @@
...
+
+* * *
+
+你或许会觉得这个故事有一点“标题党”。 **实际的解决方案其实真的很简单，但我花了大量的篇幅在转折上来解释为什么 React 最终选择了这套方案，以及还有哪些候选方案。**
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
你或许会觉得这个故事有一点「标题党」。 **实际的解决方案其实真的很简单，但我花了大量的篇幅在转折上来解释为什么 React 最终选择了这套方案，以及还有哪些候选方案。**
```

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md [hunk 35]
```diff
@@ -0,0 +1,549 @@
...
+
+以我的经验来看，设计一个库的 API 也经常会遇到这种情况。为了一个 API 能够简单易用，你经常需要考虑语义化（可能的话，为多种语言考虑，包括未来的发展方向）、运行时性能、有或没有编译时步骤的工程效能、生态的状态以及打包方案、早期的警告，以及很多其它问题。最终的结果未必总是最优雅的，但必须要是可用的。
+
+**如果最终的 API 成功的话, _它的用户_ 永远不必思考这一过程**。他们只需要专心创建应用就好了。
```
Comment (chloewlin) [inline src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]:
```suggestion
**如果最终的 API 成功的话，_它的用户_ 永远不必思考这一过程**。他们只需要专心创建应用就好了。
```

**Comment by chloewlin** [comment]
Hi @tonghuashuo thank you for submitting this translation! I added some suggestions for your references. Most of them are about missing commas or periods, removing whitespaces and using full-width punctuations based on this [guideline](https://github.com/mzlogin/chinese-copywriting-guidelines#%E5%85%A8%E8%A7%92%E5%92%8C%E5%8D%8A%E8%A7%92). Let me know what you think! Once you're done fixing them, I can help merging this in.

File: src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md (hunk 9 above)
Comment (tonghuashuo) [reply src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]: I think that a [space between links](https://github.com/mzlogin/chinese-copywriting-guidelines#%E9%93%BE%E6%8E%A5%E4%B9%8B%E9%97%B4%E5%A2%9E%E5%8A%A0%E7%A9%BA%E6%A0%BC)

Comment (chloewlin) [reply src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hans.md]: @tonghuashuo Ah, so I checked and looks like you can indeed choose between adding a space or not. It says it's controversial. In this case, you can skip my suggestions on space before links then.

**Comment by tonghuashuo** [comment]
Thanks for the review @chloewlin . That's a lot of work. I've updated the translation.

As for the [Space between links](https://github.com/mzlogin/chinese-copywriting-guidelines#%E9%93%BE%E6%8E%A5%E4%B9%8B%E9%97%B4%E5%A2%9E%E5%8A%A0%E7%A9%BA%E6%A0%BC) thing. Turned out it's under 'controversial'. Personally I prefer a space in between to show some difference. But I'm cool with the community preferring otherwise.

**Comment by xueqingxiao** [comment]
Hi @chloewlin ， I've done some reviewer's work. if you feels that is not making sense, ignore it.

**Comment by chloewlin** [comment]
Hi @xueqingxiao, I am not sure what you mean. Do you mean that you've added some reviews to this particular PR? 

@tonghuashuo thank you so much for adopting these suggestions. I think your PR is ready to be merged. If @xueqingxiao is okay, I am going to merge this in soon.

**Comment by xueqingxiao** [comment]
@chloewlin I have pinned wrong person, sorry for this confused. if @tonghuashuo can accept these comments should be great.
@chloewlin can you set me as a reviewer for each Chinese simplified translation. I can help to review grammar, spell and meaning.

**Comment by chloewlin** [comment]
Hi @xueqingxiao, you don't have to a reviewer to review. You can just review and add suggestions as you see fit.

**Pull request merged**